File: python/robot-name/robot_name.py
from string import ascii_uppercase, digits
from random import choices


past_names: set[str] = set()
max_names: int = 26**2 * 10**3


class Robot:

    name: str

    def __init__(self) -> None:
        self.reset()

    def reset(self) -> None:
        assert len(past_names) < max_names
        while (new := self._generate_name()) in past_names:
            pass
        past_names.add(new)
        self.name = new

    @staticmethod
    def _generate_name() -> str:
        return ''.join(choices(ascii_uppercase, k=2) + choices(digits, k=3))

File: kotlin/src/main/kotlin/robot_name/RobotName.kt
package robot_name

class Robot {

    val name: String get() = _name
    private var _name: String = generateName()

    tailrec fun reset(new: String = generateName()) {
        if (new == _name) return reset()
        cache.remove(_name)
        _name = new
    }

    companion object {
        private val cache = mutableSetOf<String>()
        private val LETTERS = 'A'..'Z'
        private val DIGITS = '0'..'9'
        private const val CACHE_MAX_SIZE = 26 * 26 * 10 * 10 * 10
        tailrec fun generateName(): String {
            check(cache.size < CACHE_MAX_SIZE) { "Max number of unique names has been generated." }
            val name = buildString {
                repeat(2) { append(LETTERS.random()) }
                repeat(3) { append(DIGITS.random()) }
            }
            return if (cache.add(name)) name else generateName()
        }
    }

}
